File: backend/app/schemas/execution.py
from typing import Any

from pydantic import BaseModel, Field


class ExecutionMarketResponse(BaseModel):
    symbol: str
    symbol_id: int
    price_precision: int
    quantity_precision: int
    step_size: str
    min_notional: str
    mark_price: str | None = None
    status: str | None = None


class OrderPreviewRequest(BaseModel):
    symbol: str = "BTC-USD"
    side: str = "long"
    size_usd: float = Field(gt=0)
    order_type: str = "market"
    market_type: str = "perps"


class OrderPreviewResponse(BaseModel):
    account_id: int
    symbol: str
    symbol_id: int
    side: str
    size_usd: float
    order_type: str
    cl_ord_id: str
    request_body: dict[str, Any]
    risk_allowed: bool
    risk_blocked_reason: str | None = None
    risk: dict[str, Any] | None = None
    signing_nonce: int
    signing_domain: dict[str, Any]
    signing_types: dict[str, Any]
    signing_message: dict[str, Any]


class SignedOrderSubmitRequest(BaseModel):
    request_body: dict[str, Any]
    api_key_name: str = Field(min_length=1, max_length=64)
    api_sign: str = Field(min_length=10)
    api_nonce: str = Field(min_length=1)
    market_type: str = "perps"


class OrderSubmitResponse(BaseModel):
    success: bool
    cl_ord_id: str | None = None
    order_id: str | None = None
    sodex_response: dict[str, Any] | None = None
    error: str | None = None
    sync_summary: dict[str, Any] | None = None

File: frontend/src/lib/api/execution.ts
import { api } from '@/lib/api/client'
import type { PreTradeRiskResponse } from '@/lib/api/risk'

export interface ExecutionMarket {
  symbol: string
  symbol_id: number
  price_precision: number
  quantity_precision: number
  step_size: string
  min_notional: string
  mark_price: string | null
  status: string | null
}

export interface OrderPreview {
  account_id: number
  symbol: string
  symbol_id: number
  side: string
  size_usd: number
  order_type: string
  cl_ord_id: string
  request_body: Record<string, unknown>
  risk_allowed: boolean
  risk_blocked_reason: string | null
  risk: PreTradeRiskResponse | null
  signing_nonce: number
  signing_domain: Record<string, unknown>
  signing_types: Record<string, { name: string; type: string }[]>
  signing_message: Record<string, unknown>
}

export interface OrderSubmitResult {
  success: boolean
  cl_ord_id: string | null
  order_id: string | null
  sodex_response: Record<string, unknown> | null
  error: string | null
  sync_summary: Record<string, unknown> | null
}

export function getExecutionMarkets(symbol?: string) {
  const query = symbol ? `?symbol=${encodeURIComponent(symbol)}` : ''
  return api.get<ExecutionMarket[]>(`/api/v1/execution/markets${query}`)
}

export function previewOrder(params: {
  symbol: string
  side: 'long' | 'short'
  size_usd: number
  order_type?: string
}) {
  return api.post<OrderPreview>('/api/v1/execution/preview', {
    symbol: params.symbol,
    side: params.side,
    size_usd: params.size_usd,
    order_type: params.order_type ?? 'market',
    market_type: 'perps',
  })
}

export function submitSignedOrder(params: {
  request_body: Record<string, unknown>
  api_key_name: string
  api_sign: string
  api_nonce: string
}) {
  return api.post<OrderSubmitResult>('/api/v1/execution/submit', {
    ...params,
    market_type: 'perps',
  })
}
